numpyro/util.py

```diff
@@ -309,3 +311,190 @@ def soft_vmap(fn, xs, batch_ndims=1, chunk_size=None):
+    plate_samples = {k: [name for name in plate_deps
+                     if k in [p[0] for p in plate_deps[name]]] for k in plate_sites}
+
+    def get_log_probs(sample, seed=0):
+        with handlers.trace() as tr, handlers.seed(model, seed), handlers.substitute(data=sample):
+            model(*args, **kwargs)
+        return {name: site['fn'].log_prob(site['value'])
+                for name, site in tr.items() if site['type'] == 'sample'}
+
+    samples = {name: site['value'] for name, site in trace.items()
+               if site['type'] == 'sample' and not site['is_observed']
+               and not site['fn'].is_discrete}
+    log_prob_grads = jax.jacobian(get_log_probs)(samples)
+    sample_deps = {}
+    for name, grads in log_prob_grads.items():
+        sample_deps[name] = {n for n in grads if n != name and (grads[n] != 0).any()}
+
+    # find discrete -> continuous dependency
+    samples = {name: site['value'] for name, site in trace.items() if site['type'] == 'sample'}
+    discrete_sites = [name for name, site in trace.items() if site['type'] == 'sample'
+                      and not site['is_observed'] and site['fn'].is_discrete]
+    log_probs_prototype = get_log_probs(samples)
+    for name in discrete_sites:
+        samples_ = samples.copy()
+        samples_.pop(name)
+        for i in range(num_tries):
+            log_probs = get_log_probs(samples_, seed=i + 1)
+            for var in samples:
+                if var == name:
+                    continue
+                if (log_probs[var] != log_probs_prototype[var]).any():
+                    sample_deps[var] |= {name}
+    sample_sample = {}
+    for name in samples:
+        sample_sample[name] = [var for var in samples if var in sample_deps[name]]
+    return {'sample_sample': sample_sample, 'sample_plate': plate_deps,
+            'plate_sample': plate_samples, 'observed': obs_sites}
+
+
+def generate_graph_specification(model_relations):
+    """
+    Convert model relations into data structure which can be readily
+    converted into a network.
+    """
+    # group nodes by plate
+    plate_groups = dict(model_relations['plate_sample'])
+    plate_rvs = {rv for rvs in plate_groups.values() for rv in rvs}
+    plate_groups[None] = [
+        rv for rv in model_relations['sample_sample'] if rv not in plate_rvs
+    ]  # RVs which are in no plate
+
+    # retain node metadata
+    node_data = {}
+    for rv in model_relations['sample_sample']:
+        node_data[rv] = {
+            'is_observed': rv in model_relations['observed'],
+        }
+
+    # infer plate structure
+    # TODO: subset relation might not always hold, raise exception if not
+    plate_data = {}
+    for plate1, plate2 in list(itertools.permutations(plate_groups, 2)):
+        if plate1 is None or plate2 is None:
+            continue
+
+        if set(plate_groups[plate1]) < set(plate_groups[plate2]):
+            plate_data[plate1] = {'parent': plate2}
```

Comment:
I assume you are referring to a model such as
```python
def plate_improper_subsets():
    with numpyro.plate('N', 10):
        with numpyro.plate('M', 10):
            numpyro.sample('x', dist.Normal(0, 1))
```

By just looking at the RV lists, we would never be able to figure out which plate is the parent and which is the child, would we?

Reply:
You are right. We are using a convention that the order follows what appears first in the trace (in case two plates have the same content). Could you add a comment here in case we look at this again in the future?